Обратная сторона реальности

RuslanPro01 opened the pull request:
---
:mortar_board: [Обратная сторона реальности](https://up.htmlacademy.ru/react/12/user/854371/tasks/7)

Commit:
refactor: Поменяет стили для page-404.tsx

- Перенесет стили в отдельный файл css
- Немного поправит семантику в коде
- Удались перечисления стилей из файла компонента

Commit:
feat: Создаст моковые данные фильмов и комментариев, созадасть моковую функцию для получения определенного колличество комметариев из списка.

Commit:
refactor: Добавит описание некоторым переменных и функциям иоковых данных

Commit:
feat: Начнет реализацию наплнения карточек с фильмами

Commit:
feat: Научи новый компонент film-cards обрисовывать карточки фильмов

Commit:
feat: Будет передавать в стейт `id` активной карточки фильма.

Commit:
feat: Научит компонент my-list.tsx отображать только фильмы из избранного

Commit:
feat: Вынесет heder и footer в отдельные компоненты

— Создаст компоненты шапки и подвала страницы.
— Подключит из на страницы

Commit:
feat: Начнет реализовывать динамическое переключение между страницами

Commit:
feat: закончит реализацию связывания страниц между собой

Commit:
feat: Вынесет форму ревью в отдельный компонент.

komaritskaya commented on project/src/components/app/app.tsx:
```diff
@@ -5,34 +5,36 @@ import Page404 from '../../pages/page-404/page-404';
+          <Route index element={<Main filmName={filmName} yearFilm={yearFilm} filmGenre={filmGenre} films={films}/>}/>
           <Route path={Login} element={<SignIn/>}/>
-          <Route path={Films.MainPage}>
+          <Route path={FilmsPages.MainPage}>
```
Здесь внутри `MoviePage` нужно будет еще реализовать навигацию между вкладками (описание, коменты и т.д.). В этом тебе поможет [Outlet](https://reactrouter.com/en/main/components/outlet)

komaritskaya commented on project/src/components/film-cards/small-film-card.tsx:
```diff
@@ -0,0 +1,34 @@
+import {useState} from 'react';
```
Я бы все-таки вернула этот файл в свою папку)

komaritskaya commented on project/src/components/film-cards/small-film-card.tsx:
```diff
@@ -0,0 +1,34 @@
+import {useState} from 'react';
+import {Link} from 'react-router-dom';
+
+type SmallFilm = {
```
Небольшое замечание по неймингу. Логично заводить `SmallFilmCard`, если у нас есть big card. Здесь это единственный вариант карточек, который есть, поэтому я бы убрала слово small из названия.
Если же этим ты хочешь показать, что мы берем не все свойства фильма, а только некоторые, то я бы предложила вообще не рассматривать props как сущность фильма. Это просто объект, в котором мы передаем информацию из одного компонента в другой и кроме полей фильма там может быть и другая инфа тоже. Т.е. лучше называть такие типа в формате `имя компонента + Props` (например, `FilmCardProps`).

Ну и еще "красивее" будет передавать в карточку не каждое поле по отдельности, а полностью весь фильм, и внутри уже его деструктурировать. Т.к. допустим если мы захотим кроме имени и картинки добавить еще, например, жанр, нам нужно будет поменять это не только в самом компоненте карточки, но еще добавить пропсы во всех местах, где эта карточка используется.

komaritskaya commented on project/src/components/film-cards/film-cards.tsx:
```diff
@@ -0,0 +1,28 @@
+}
+
+function FilmCards({films, onlyFavorite}: PropsFilmCard): JSX.Element {
+  if (onlyFavorite) {
```
Во-первых, здесь основным DOM-элементом будет не фрагмент, а `<div className="catalog__films-list">` (обрати внимание, везде, где вставляются карточки, есть этот контейнер).

Во-вторых, здесь тебе не нужно свойство `onlyFavorite`. Список фильмов должен принимать только массив из фильмов, а уже как этот массив сформирован - ему не должно быть интересно, вся фильтрация должна происходить в родительском компоненте.

komaritskaya commented on project/src/components/film-cards/small-film-card.tsx:
```diff
@@ -0,0 +1,34 @@
+}
+
+function SmallFilmCard({name, previewImage, id}: SmallFilm): JSX.Element {
+  const [activeFilm, setActiveFilm] = useState<{id: number} | null>(null);
```
Тут я бы немного по-другому разделила ответственность - стейт с айдишником активного фильма лучше хранить в компоненте списка (это даже предлагается в задании), а в саму карточку передавать хэндлеры и булево свойство `isActive`. При этом ты в стейт записываешь объект с полем `id`, но сам объект тут - лишняя обертка, достаточно просто записывать айдишник.

Commit:
feat: Реализует навигацию по табам на странице фильма

- Чтобы реализовать сохранение состояния по ссылке, добавил новые пути в перечислении и компоненте `App`
- Вынес разметку табов в отдельные компоненты `details.tsx`, `overview.tsx`, `review.tsx`. Также в отдельный компонент вынес меню навигации по табам

Commit:
refactor: Исправит пути в компонентах `page-404.tsx` и `player.tsx`.

Commit:
refactor: Внесет исправления по замечаниям наставника

- Сделает нейминг компонентов более понятным
- Вынесет компонент `film-card.tsx` в отдельную папку. Вынесет логику отрисовки карточек в родительские компоненты.
- Упростить карточку с фильмом, перенеся ответственность за определение активного фильма в родительский компонент
- Сделает автоматическую навигацию к основному табу на странице фильма, если пользователь внес в строку только фильм

komaritskaya commented on project/src/components/footer.tsx:
```diff
@@ -0,0 +1,21 @@
+import {Link} from 'react-router-dom';
```
Давай будем соблюдать общую структуру проекта и обернем этот компонент тоже в папку.

komaritskaya commented on project/src/components/film-cards/film-cards.tsx:
```diff
@@ -0,0 +1,44 @@
+              previewImage={previewImage}
+              id={id}
+              key={id}
+              isActiveFilm={activeFilm.isActiveFilm}
```
Я предлагаю убрать из стейта лишнее поле `isActiveFilm` и оставить только айдишник активного фильма. Тогда мы сможем здесь прописать условие как `isActiveFilm={id === activeFilm}` и не хранить в компоненте лишнюю  информацию.

komaritskaya commented on project/src/pages/movie-page/movie-page.tsx:
```diff
@@ -1,4 +1,28 @@
+
+  const location = useLocation();
+  useEffect(() => {
+    const isOverviewActive = location.pathname.endsWith(Path.FilmsPages.Tabs.Overview);
```
Здесь можно не делать отдельную вкладку для `overview`, а сделать ее вкладкой по умолчанию, т.е. путь к ней будет не `/films/:id/overview`, а просто `/films/:id`. Тогда можно будет не мучаться с `useEffect` и навигацией.

Commit:
refactor: Вторая доработка по замечаниям наставника

- Перенесет footer.tsx в отдельную директорию.
- Избавится от принудительного роутинга на таб Overview, сделает записьв Outlet по умолчанию при переходе к карточке фильма из адресной строки
- Уберет сценарий eject из package.json
